## Answer the following questions using the wine review dataset from previous classd
1.Set the index of the dataframe to country
2.Using the index and any country of your choice, show the reviews for those country 
Sample code
df.loc[df.country == 'Italy']
can you glean any insight from this? Perhaps the percentage of wines from the selected country.
3.Supposing wines are reviewed on a 80-to-100 point scale, and wines with greater than 90 points are average, can you check what percentage of the total wines are above average? What about the percentage in the country you selected from question 2?
4.Check the pandas ‘isin’ operator and use it to perform any operation of your choice on the dataset
5.Suppose we'll buy any wine that's made in Italy or which is rated above average. Can you select or show these wines using the or (|) operator?
6.Read about the notnull operator and use it to check for wines whose prices are not null
7.Please share something you learnt outside of class on thursday


In [14]:
import pandas as pd

In [15]:
Wine_review = pd.read_csv ('Wine_review.csv') 

In [16]:
Wine_review

Unnamed: 0 country                                        description  \
0                0      US  This tremendous 100% varietal wine hails from ...   
1                1   Spain  Ripe aromas of fig, blackberry and cassis are ...   
2                2      US  Mac Watson honors the memory of a wine once ma...   
3                3      US  This spent 20 months in 30% new French oak, an...   
4                4  France  This is the top wine from La Bégude, named aft...   
...            ...     ...                                                ...   
150925      150925   Italy  Many people feel Fiano represents southern Ita...   
150926      150926  France  Offers an intriguing nose with ginger, lime an...   
150927      150927   Italy  This classic example comes from a cru vineyard...   
150928      150928  France  A perfect salmon shade, with scents of peaches...   
150929      150929   Italy  More Pinot Grigios should taste like this. A r...   

                                 designation  points  price  \
0                          Martha's Vineyard      96  235.0   
1       Carodorum Selección Especial Reserva      96  110.0   
2              Special Selected Late Harvest      96   90.0   
3                                    Reserve      96   65.0   
4                                 La Brûlade      95   66.0   
...                                      ...     ...    ...   
150925                                   NaN      91   20.0   
150926                        Cuvée Prestige      91   27.0   
150927                         Terre di Dora      91   20.0   
150928                       Grand Brut Rosé      90   52.0   
150929                                   NaN      90   15.0   

                  province           region_1           region_2  \
0               California        Napa Valley               Napa   
1           Northern Spain               Toro                NaN   
2               California     Knights Valley             Sonoma   
3                   Oregon  Willamette Valley  Willamette Valley   
4                 Provence             Bandol                NaN   
...                    ...                ...                ...   
150925      Southern Italy  Fiano di Avellino                NaN   
150926           Champagne          Champagne                NaN   
150927      Southern Italy  Fiano di Avellino                NaN   
150928           Champagne          Champagne                NaN   
150929  Northeastern Italy         Alto Adige                NaN   

                   variety                   winery  
0       Cabernet Sauvignon                    Heitz  
1            Tinta de Toro  Bodega Carmen Rodríguez  
2          Sauvignon Blanc                 Macauley  
3               Pinot Noir                    Ponzi  
4       Provence red blend     Domaine de la Bégude  
...                    ...                      ...  
150925         White Blend    Feudi di San Gregorio  
150926     Champagne Blend                H.Germain  
150927         White Blend                Terredora  
150928     Champagne Blend                   Gosset  
150929        Pinot Grigio            Alois Lageder  

[150930 rows x 11 columns]

In [18]:
df.set_index ('country')

Unnamed: 0                                        description  \
country                                                                  
US                0  This tremendous 100% varietal wine hails from ...   
Spain             1  Ripe aromas of fig, blackberry and cassis are ...   
US                2  Mac Watson honors the memory of a wine once ma...   
US                3  This spent 20 months in 30% new French oak, an...   
France            4  This is the top wine from La Bégude, named aft...   
...             ...                                                ...   
Italy        150925  Many people feel Fiano represents southern Ita...   
France       150926  Offers an intriguing nose with ginger, lime an...   
Italy        150927  This classic example comes from a cru vineyard...   
France       150928  A perfect salmon shade, with scents of peaches...   
Italy        150929  More Pinot Grigios should taste like this. A r...   

                                  designation  points  price  \
country                                                        
US                          Martha's Vineyard      96  235.0   
Spain    Carodorum Selección Especial Reserva      96  110.0   
US              Special Selected Late Harvest      96   90.0   
US                                    Reserve      96   65.0   
France                             La Brûlade      95   66.0   
...                                       ...     ...    ...   
Italy                                     NaN      91   20.0   
France                         Cuvée Prestige      91   27.0   
Italy                           Terre di Dora      91   20.0   
France                        Grand Brut Rosé      90   52.0   
Italy                                     NaN      90   15.0   

                   province           region_1           region_2  \
country                                                             
US               California        Napa Valley               Napa   
Spain        Northern Spain               Toro                NaN   
US               California     Knights Valley             Sonoma   
US                   Oregon  Willamette Valley  Willamette Valley   
France             Provence             Bandol                NaN   
...                     ...                ...                ...   
Italy        Southern Italy  Fiano di Avellino                NaN   
France            Champagne          Champagne                NaN   
Italy        Southern Italy  Fiano di Avellino                NaN   
France            Champagne          Champagne                NaN   
Italy    Northeastern Italy         Alto Adige                NaN   

                    variety                   winery  
country                                               
US       Cabernet Sauvignon                    Heitz  
Spain         Tinta de Toro  Bodega Carmen Rodríguez  
US          Sauvignon Blanc                 Macauley  
US               Pinot Noir                    Ponzi  
France   Provence red blend     Domaine de la Bégude  
...                     ...                      ...  
Italy           White Blend    Feudi di San Gregorio  
France      Champagne Blend                H.Germain  
Italy           White Blend                Terredora  
France      Champagne Blend                   Gosset  
Italy          Pinot Grigio            Alois Lageder  

[150930 rows x 10 columns]

In [20]:
df.loc[df.country == 'US']

Unnamed: 0 country                                        description  \
0                0      US  This tremendous 100% varietal wine hails from ...   
2                2      US  Mac Watson honors the memory of a wine once ma...   
3                3      US  This spent 20 months in 30% new French oak, an...   
8                8      US  This re-named vineyard was formerly bottled as...   
9                9      US  The producer sources from two blocks of the vi...   
...            ...     ...                                                ...   
150892      150892      US  A light, earthy wine, with violet, berry and t...   
150896      150896      US  Some raspberry fruit in the aroma, but things ...   
150914      150914      US  Old-gold in color, and thick and syrupy. The a...   
150915      150915      US  Decades ago, Beringer’s then-winemaker Myron N...   
150916      150916      US  An impressive wine that presents a full bouque...   

                          designation  points  price    province  \
0                   Martha's Vineyard      96  235.0  California   
2       Special Selected Late Harvest      96   90.0  California   
3                             Reserve      96   65.0      Oregon   
8                              Silice      95   65.0      Oregon   
9                Gap's Crown Vineyard      95   60.0  California   
...                               ...     ...    ...         ...   
150892                        Coastal      82   10.0  California   
150896                            NaN      82   10.0  California   
150914    Late Harvest Cluster Select      94   25.0  California   
150915                    Nightingale      93   30.0  California   
150916                      J. Schram      93   65.0  California   

                  region_1                 region_2             variety  \
0              Napa Valley                     Napa  Cabernet Sauvignon   
2           Knights Valley                   Sonoma     Sauvignon Blanc   
3        Willamette Valley        Willamette Valley          Pinot Noir   
8       Chehalem Mountains        Willamette Valley          Pinot Noir   
9             Sonoma Coast                   Sonoma          Pinot Noir   
...                    ...                      ...                 ...   
150892          California         California Other              Merlot   
150896          California         California Other          Pinot Noir   
150914     Anderson Valley  Mendocino/Lake Counties      White Riesling   
150915         North Coast              North Coast         White Blend   
150916         Napa Valley                     Napa     Champagne Blend   

             winery  
0             Heitz  
2          Macauley  
3             Ponzi  
8         Bergström  
9         Blue Farm  
...             ...  
150892     Callaway  
150896      Camelot  
150914      Navarro  
150915     Beringer  
150916  Schramsberg  

[62397 rows x 11 columns]

In [22]:
df.loc[review.country == 'US']

Unnamed: 0 country                                        description  \
0                0      US  This tremendous 100% varietal wine hails from ...   
2                2      US  Mac Watson honors the memory of a wine once ma...   
3                3      US  This spent 20 months in 30% new French oak, an...   
8                8      US  This re-named vineyard was formerly bottled as...   
9                9      US  The producer sources from two blocks of the vi...   
...            ...     ...                                                ...   
150892      150892      US  A light, earthy wine, with violet, berry and t...   
150896      150896      US  Some raspberry fruit in the aroma, but things ...   
150914      150914      US  Old-gold in color, and thick and syrupy. The a...   
150915      150915      US  Decades ago, Beringer’s then-winemaker Myron N...   
150916      150916      US  An impressive wine that presents a full bouque...   

                          designation  points  price    province  \
0                   Martha's Vineyard      96  235.0  California   
2       Special Selected Late Harvest      96   90.0  California   
3                             Reserve      96   65.0      Oregon   
8                              Silice      95   65.0      Oregon   
9                Gap's Crown Vineyard      95   60.0  California   
...                               ...     ...    ...         ...   
150892                        Coastal      82   10.0  California   
150896                            NaN      82   10.0  California   
150914    Late Harvest Cluster Select      94   25.0  California   
150915                    Nightingale      93   30.0  California   
150916                      J. Schram      93   65.0  California   

                  region_1                 region_2             variety  \
0              Napa Valley                     Napa  Cabernet Sauvignon   
2           Knights Valley                   Sonoma     Sauvignon Blanc   
3        Willamette Valley        Willamette Valley          Pinot Noir   
8       Chehalem Mountains        Willamette Valley          Pinot Noir   
9             Sonoma Coast                   Sonoma          Pinot Noir   
...                    ...                      ...                 ...   
150892          California         California Other              Merlot   
150896          California         California Other          Pinot Noir   
150914     Anderson Valley  Mendocino/Lake Counties      White Riesling   
150915         North Coast              North Coast         White Blend   
150916         Napa Valley                     Napa     Champagne Blend   

             winery  
0             Heitz  
2          Macauley  
3             Ponzi  
8         Bergström  
9         Blue Farm  
...             ...  
150892     Callaway  
150896      Camelot  
150914      Navarro  
150915     Beringer  
150916  Schramsberg  

[62397 rows x 11 columns]

Out of the 150,930 rows, 62,397 rows contains wine made in the US
and that accounts for approximately 41% of the total wine.


In [23]:
df.loc[(review.country == 'US')&(review.points > 90)]

Unnamed: 0 country                                        description  \
0                0      US  This tremendous 100% varietal wine hails from ...   
2                2      US  Mac Watson honors the memory of a wine once ma...   
3                3      US  This spent 20 months in 30% new French oak, an...   
8                8      US  This re-named vineyard was formerly bottled as...   
9                9      US  The producer sources from two blocks of the vi...   
...            ...     ...                                                ...   
150778      150778      US  Deep, layered aromas of smoky honey, lees and ...   
150780      150780      US  Very much in the vein of a super-oaked, ripe C...   
150914      150914      US  Old-gold in color, and thick and syrupy. The a...   
150915      150915      US  Decades ago, Beringer’s then-winemaker Myron N...   
150916      150916      US  An impressive wine that presents a full bouque...   

                          designation  points  price    province  \
0                   Martha's Vineyard      96  235.0  California   
2       Special Selected Late Harvest      96   90.0  California   
3                             Reserve      96   65.0      Oregon   
8                              Silice      95   65.0      Oregon   
9                Gap's Crown Vineyard      95   60.0  California   
...                               ...     ...    ...         ...   
150778                            NaN      91   22.0  California   
150780                            NaN      91   24.0  California   
150914    Late Harvest Cluster Select      94   25.0  California   
150915                    Nightingale      93   30.0  California   
150916                      J. Schram      93   65.0  California   

                    region_1                 region_2             variety  \
0                Napa Valley                     Napa  Cabernet Sauvignon   
2             Knights Valley                   Sonoma     Sauvignon Blanc   
3          Willamette Valley        Willamette Valley          Pinot Noir   
8         Chehalem Mountains        Willamette Valley          Pinot Noir   
9               Sonoma Coast                   Sonoma          Pinot Noir   
...                      ...                      ...                 ...   
150778  Arroyo Grande Valley            Central Coast          Chardonnay   
150780               Chalone            Central Coast         Pinot Blanc   
150914       Anderson Valley  Mendocino/Lake Counties      White Riesling   
150915           North Coast              North Coast         White Blend   
150916           Napa Valley                     Napa     Champagne Blend   

                  winery  
0                  Heitz  
2               Macauley  
3                  Ponzi  
8              Bergström  
9              Blue Farm  
...                  ...  
150778            Talley  
150780  Chalone Vineyard  
150914           Navarro  
150915          Beringer  
150916       Schramsberg  

[14427 rows x 11 columns]

In [28]:
df.loc[review.country.isin(['Argentina','France','US'])]
df.loc[review.points.isin([94, 92, 91])]
df.loc[review.price.isin([30.0,25.0,24.0])]

Unnamed: 0    country  \
20              20         US   
45              45      Italy   
57              57         US   
58              58         US   
60              60  Argentina   
...            ...        ...   
150913      150913     France   
150914      150914         US   
150915      150915         US   
150917      150917     France   
150923      150923     France   

                                              description  \
20      Heitz has made this stellar rosé from the rare...   
45      A blend of 90% Sangiovese and 10% Canaiolo, th...   
57      While exuberantly fruity, almost tropical on t...   
58      Cabernet makes up just over half of this blend...   
60      Concentrated, ripe blackberry and cassis aroma...   
...                                                   ...   
150913  Inky and rustic, yet in a refined manner. This...   
150914  Old-gold in color, and thick and syrupy. The a...   
150915  Decades ago, Beringer’s then-winemaker Myron N...   
150917  Light and elegant, this spicy, lively wine is ...   
150923  Rich and toasty, with tiny bubbles. The bouque...   

                                     designation  points  price  \
20                                    Grignolino      95   24.0   
45               Vigneto Odoardo Beccari Riserva      90   30.0   
57                                       Reserve      90   25.0   
58                                   Final Final      90   30.0   
60      The Apple Doesn't Fall Far From The Tree      91   30.0   
...                                          ...     ...    ...   
150913                                       NaN      94   30.0   
150914               Late Harvest Cluster Select      94   25.0   
150915                               Nightingale      93   30.0   
150917                             Brut Mosaïque      92   30.0   
150923                                  Demi-Sec      91   30.0   

                province              region_1                 region_2  \
20            California           Napa Valley                     Napa   
45               Tuscany      Chianti Classico                      NaN   
57              New York          Finger Lakes             Finger Lakes   
58            Washington  Columbia Valley (WA)          Columbia Valley   
60      Mendoza Province               Mendoza                      NaN   
...                  ...                   ...                      ...   
150913      Rhône Valley   Châteauneuf-du-Pape                      NaN   
150914        California       Anderson Valley  Mendocino/Lake Counties   
150915        California           North Coast              North Coast   
150917         Champagne             Champagne                      NaN   
150923         Champagne             Champagne                      NaN   

                         variety                winery  
20                          Rosé                 Heitz  
45                     Red Blend          Vignavecchia  
57                      Riesling  Dr. Konstantin Frank  
58      Cabernet Sauvignon-Syrah                Efeste  
60                        Malbec     Matias Riccitelli  
...                          ...                   ...  
150913     Rhône-style Red Blend       Le Vieux Donjon  
150914            White Riesling               Navarro  
150915               White Blend              Beringer  
150917           Champagne Blend              Jacquart  
150923           Champagne Blend              Jacquart  

[14847 rows x 11 columns]

In [29]:
df.loc[(review.designation.notnull())]
df.loc[(review.points.notnull())]

Unnamed: 0 country                                        description  \
0                0      US  This tremendous 100% varietal wine hails from ...   
1                1   Spain  Ripe aromas of fig, blackberry and cassis are ...   
2                2      US  Mac Watson honors the memory of a wine once ma...   
3                3      US  This spent 20 months in 30% new French oak, an...   
4                4  France  This is the top wine from La Bégude, named aft...   
...            ...     ...                                                ...   
150925      150925   Italy  Many people feel Fiano represents southern Ita...   
150926      150926  France  Offers an intriguing nose with ginger, lime an...   
150927      150927   Italy  This classic example comes from a cru vineyard...   
150928      150928  France  A perfect salmon shade, with scents of peaches...   
150929      150929   Italy  More Pinot Grigios should taste like this. A r...   

                                 designation  points  price  \
0                          Martha's Vineyard      96  235.0   
1       Carodorum Selección Especial Reserva      96  110.0   
2              Special Selected Late Harvest      96   90.0   
3                                    Reserve      96   65.0   
4                                 La Brûlade      95   66.0   
...                                      ...     ...    ...   
150925                                   NaN      91   20.0   
150926                        Cuvée Prestige      91   27.0   
150927                         Terre di Dora      91   20.0   
150928                       Grand Brut Rosé      90   52.0   
150929                                   NaN      90   15.0   

                  province           region_1           region_2  \
0               California        Napa Valley               Napa   
1           Northern Spain               Toro                NaN   
2               California     Knights Valley             Sonoma   
3                   Oregon  Willamette Valley  Willamette Valley   
4                 Provence             Bandol                NaN   
...                    ...                ...                ...   
150925      Southern Italy  Fiano di Avellino                NaN   
150926           Champagne          Champagne                NaN   
150927      Southern Italy  Fiano di Avellino                NaN   
150928           Champagne          Champagne                NaN   
150929  Northeastern Italy         Alto Adige                NaN   

                   variety                   winery  
0       Cabernet Sauvignon                    Heitz  
1            Tinta de Toro  Bodega Carmen Rodríguez  
2          Sauvignon Blanc                 Macauley  
3               Pinot Noir                    Ponzi  
4       Provence red blend     Domaine de la Bégude  
...                    ...                      ...  
150925         White Blend    Feudi di San Gregorio  
150926     Champagne Blend                H.Germain  
150927         White Blend                Terredora  
150928     Champagne Blend                   Gosset  
150929        Pinot Grigio            Alois Lageder  

[150930 rows x 11 columns]

In [31]:
df.loc[review.price.isnull()]

Unnamed: 0      country  \
32              32        Italy   
56              56       France   
72              72        Italy   
82              82        Italy   
116            116        Spain   
...            ...          ...   
150377      150377  New Zealand   
150378      150378  New Zealand   
150587      150587       Canada   
150673      150673           US   
150922      150922        Italy   

                                              description       designation  \
32      Underbrush, scorched earth, menthol and plum s...     Vigna Piaggia   
56      Delicious while also young and textured, this ...           Le Pavé   
72      This offers aromas of red rose, wild berry, da...    Bussia Riserva   
82      Berry, baking spice, dried iris, mint and a hi...  Palliano Riserva   
116     Aromas of brandied cherry and crème de cassis ...       Dulce Tinto   
...                                                   ...               ...   
150377  Light and a bit herbal, like a pleasant St.-Jo...          Matheson   
150378  Impressive purple color, but less intense on t...               NaN   
150587  Shows pronounced oily, earthy, almost tobacco-...           Icewine   
150673  Cherry-scented, clean and fruity. Good concent...               NaN   
150922  Made by 30-ish Roberta Borghese high above Man...         Superiore   

        points  price            province                    region_1  \
32          90    NaN             Tuscany      Brunello di Montalcino   
56          90    NaN        Loire Valley                    Sancerre   
72          91    NaN            Piedmont                      Barolo   
82          91    NaN            Piedmont                       Roero   
116         86    NaN             Levante                     Jumilla   
...        ...    ...                 ...                         ...   
150377      84    NaN         Hawke's Bay                         NaN   
150378      84    NaN       Martinborough                         NaN   
150587      90    NaN             Ontario       Lake Erie North Shore   
150673      87    NaN          California            Dry Creek Valley   
150922      91    NaN  Northeastern Italy  Colli Orientali del Friuli   

       region_2          variety             winery  
32          NaN       Sangiovese    Abbadia Ardenga  
56          NaN  Sauvignon Blanc   Domaine Vacheron  
72          NaN         Nebbiolo    Silvano Bolmida  
82          NaN         Nebbiolo              Ceste  
116         NaN       Monastrell  Casa de la Ermita  
...         ...              ...                ...  
150377      NaN            Syrah       Matua Valley  
150378      NaN            Syrah             Kusuda  
150587      NaN         Riesling              Colio  
150673   Sonoma        Zinfandel        Taft Street  
150922      NaN            Tocai  Ronchi di Manzano  

[13695 rows x 11 columns]

Data cleaning means fixing bad data in your data set.

Bad data could be:

Empty cells
Data in wrong format
Wrong data
Duplicates

## Class Pandas- Groupby

In [ ]:
review.groupby()

